bug: type safety violation - nil assignable to typed variables

## Description
`arrays.first()` on an empty array returns `nil`, which can be assigned to a typed variable (e.g., `temp x int`). This violates type safety because the variable is declared as `int` but holds `nil`, causing runtime errors when used.

## Expected Behavior
Either:
1. `arrays.first()` on empty array should throw a runtime error immediately, OR
2. The return type should be explicitly nullable and require nil-checking before assignment to non-nullable types, OR
3. The compiler should reject assignment of possibly-nil values to non-nullable types

## Reproduction
```ez
import @std, @arrays
using std

do main() {
    temp empty [int] = {}
    temp first_val int = arrays.first(empty)  // No compile error!
    println("Result: ${first_val}")           // Prints "nil"
    temp result int = first_val + 1           // Runtime error!
}
```

## Actual Behavior
```
=== Testing arrays.first() on empty array ===
Result: nil
Testing nil + 1:
error[E5006]: nil reference: cannot use nil with operator '+'
```

## Why This Is A Bug
- Violates type safety: an `int` variable should never contain `nil`
- The compiler allows this without warning
- Runtime error is delayed until the value is used, making debugging harder
- Same issue likely affects: `arrays.last()`, `arrays.pop()`, map operations, etc.
- Proper type systems either disallow nil in non-nullable types or require explicit handling

--- a/pkg/typechecker/typechecker.go
+++ b/pkg/typechecker/typechecker.go
@@ -1796,6 +1796,18 @@ func (tc *TypeChecker) isMapType(typeName string) bool {
 	return strings.HasPrefix(typeName, "map[") && strings.HasSuffix(typeName, "]")
 }
 
+// isNullableType checks if a type can accept nil values
+// Only user-defined struct types can be nil in EZ
+// Arrays, maps, and primitives cannot be nil
+func (tc *TypeChecker) isNullableType(typeName string) bool {
+	// User-defined struct types can be nil
+	if t, exists := tc.types[typeName]; exists && t.Kind == StructType {
+		return true
+	}
+	// Everything else (arrays, maps, primitives, enums) cannot be nil
+	return false
+}
+
 // isHashableType checks if a type can be used as a map key
 func (tc *TypeChecker) isHashableType(typeName string) bool {
 	switch typeName {
@@ -2445,9 +2457,10 @@ func (tc *TypeChecker) typesCompatible(declared, actual string) bool {
 		return true
 	}
 
-	// nil is compatible with any reference type
+	// nil is only compatible with reference types (arrays, maps, structs)
+	// Primitive types (int, float, string, bool, char, byte) cannot be nil
 	if actual == "nil" {
-		return true
+		return tc.isNullableType(declared)
 	}
 
 	// Handle array type compatibility

--- a/pkg/typechecker/typechecker_test.go
+++ b/pkg/typechecker/typechecker_test.go
@@ -1575,6 +1575,31 @@ do main() {
 	assertHasError(t, tc, errors.E3025) // enum members must all have the same type
 }
 
+func TestNilAssignmentToPrimitiveError(t *testing.T) {
+	input := `
+do main() {
+	temp x int = nil
+}
+`
+	tc := typecheck(t, input)
+	assertHasError(t, tc, errors.E3001) // cannot assign nil to int
+}
+
+func TestNilAssignmentToStructAllowed(t *testing.T) {
+	input := `
+const Point struct {
+	x int
+	y int
+}
+
+do main() {
+	temp p Point = nil
+}
+`
+	tc := typecheck(t, input)
+	assertNoErrors(t, tc)
+}
+
 // ============================================================================
 // Global Variable Declaration Tests
 // ============================================================================

